src/test/java/io/nats/client/impl/AuthAndConnectTests.java

```diff
@@ -50,4 +56,51 @@ public void testConnectWhenClosed() throws Exception {
             assertClosed(nc);
         }
     }
+
+    /**
+     * Simulates an issue where {@link NatsConnection#closeSocket} is simultaneously called from multiple threads,
+     * and the {@link NatsConnection#reconnect} closes the connection due to overwriting the disconnecting state.
+     */
+    @RepeatedTest(5)
+    public void testNoCloseOnSimultaneouslyClosingSocket() throws Exception {
+        try (NatsTestServer ts = new NatsTestServer(false)) {
+            Options options = Options.builder()
+                    .server(ts.getURI())
+                    .maxReconnects(-1)
+                    .build();
+
+            NatsConnection nc = (NatsConnection) standardConnection(options);
+
+            // After we've connected, shut down, so we can attempt reconnecting.
+            ts.shutdown(true);
+
+            final AtomicBoolean running = new AtomicBoolean(true);
+            Thread closeSocketThread = new Thread(() -> {
+                try {
+                    nc.closeSocket(true);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+
+            Thread parallelCommunicationIssues = new Thread(() -> {
+                while (running.get()) {
+                    nc.handleCommunicationIssue(new Exception());
+                }
+            });
+
+            closeSocketThread.start();
+
+            // Ensure the closeSocket thread runs first
+            Thread.sleep(100);
+
+            parallelCommunicationIssues.start();
+
+            // Wait for some time to allow for reconnection logic to run.
+            Thread.sleep(2000);
+            running.set(false);
+
+            assertNotEquals(Connection.Status.CLOSED, nc.getStatus());
+        }
+    }
 }
```

Comment:
nit please add a blank line at the end of file

Reply:
Done